File: design_pattern/adapter.ts
// ============================================================
// Adapter Pattern — 让不兼容的接口可以一起工作
// 对比 Rust: 06_adapter.rs
// 运行: npx ts-node 06_adapter.ts
// ============================================================

// 我们期望的新接口
interface PaymentProcessor {
  pay(amountCents: number): boolean;
  refund(amountCents: number): boolean;
}

// 旧的第三方 API（无法修改）
class LegacyApi {
  constructor(private merchantId: string) {}

  processPayment(amount: number, currency: string): number {
    console.log(`[LegacyApi] merchant=${this.merchantId} pay ${amount.toFixed(2)} ${currency}`);
    return 0;
  }

  processRefund(amount: number, currency: string): number {
    console.log(`[LegacyApi] merchant=${this.merchantId} refund ${amount.toFixed(2)} ${currency}`);
    return 0;
  }
}

// 适配器
class LegacyAdapter implements PaymentProcessor {
  private api: LegacyApi;

  constructor(merchantId: string, private currency: string) {
    this.api = new LegacyApi(merchantId);
  }

  pay(amountCents: number): boolean {
    return this.api.processPayment(amountCents / 100, this.currency) === 0;
  }

  refund(amountCents: number): boolean {
    return this.api.processRefund(amountCents / 100, this.currency) === 0;
  }
}

function checkout(processor: PaymentProcessor, amountCents: number) {
  console.log(`Checkout ${amountCents} cents =>`);
  const ok = processor.pay(amountCents);
  console.log("Result:", ok ? "success" : "failed", "\n");
}

// --- main ---
console.log("=== Adapter Pattern ===\n");

const processor = new LegacyAdapter("merchant-001", "USD");
checkout(processor, 1999);
checkout(processor, 5000);

console.log("Refund $19.99:");
processor.refund(1999);

File: design_pattern/adapter.rs
// ============================================================
// Adapter Pattern — 让不兼容的接口可以一起工作
// 对比 TS: 06_adapter.ts
// 运行: cargo run --bin adapter
// ============================================================

// 我们期望的新接口
trait PaymentProcessor {
    fn pay(&self, amount_cents: u64) -> bool;
    fn refund(&self, amount_cents: u64) -> bool;
}

// 旧的第三方 API（无法修改）
struct LegacyApi {
    merchant_id: String,
}

impl LegacyApi {
    fn new(id: &str) -> Self { Self { merchant_id: id.into() } }

    fn process_payment(&self, amount: f64, currency: &str) -> i32 {
        println!("[LegacyApi] merchant={} pay {:.2} {}", self.merchant_id, amount, currency);
        0
    }

    fn process_refund(&self, amount: f64, currency: &str) -> i32 {
        println!("[LegacyApi] merchant={} refund {:.2} {}", self.merchant_id, amount, currency);
        0
    }
}

// 适配器：把旧 API 包装成新接口
struct LegacyAdapter {
    api: LegacyApi,
    currency: String,
}

impl LegacyAdapter {
    fn new(merchant_id: &str, currency: &str) -> Self {
        Self { api: LegacyApi::new(merchant_id), currency: currency.into() }
    }
}

impl PaymentProcessor for LegacyAdapter {
    fn pay(&self, amount_cents: u64) -> bool {
        self.api.process_payment(amount_cents as f64 / 100.0, &self.currency) == 0
    }
    fn refund(&self, amount_cents: u64) -> bool {
        self.api.process_refund(amount_cents as f64 / 100.0, &self.currency) == 0
    }
}

// 业务代码只依赖新接口
fn checkout(processor: &dyn PaymentProcessor, amount_cents: u64) {
    println!("Checkout {} cents =>", amount_cents);
    let ok = processor.pay(amount_cents);
    println!("Result: {}\n", if ok { "success" } else { "failed" });
}

fn main() {
    println!("=== Adapter Pattern ===\n");

    let processor = LegacyAdapter::new("merchant-001", "USD");
    checkout(&processor, 1999);
    checkout(&processor, 5000);

    println!("Refund $19.99:");
    processor.refund(1999);
}
